### PR TITLE
Draft: Introduce support for DPDK secondary mode for workers runmode and DPDK Prefilter

### PR DESCRIPTION
**Pull request by lukashino**
- [X] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [X] I have signed the Open Information Security Foundation contribution agreement at https://suricata.io/about/contribution-agreement/
- [X] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine ticket about Suricata secondary process support](https://redmine.openinfosecfoundation.org/issues/4910).
Link to [redmine ticket about prefilter application](https://redmine.openinfosecfoundation.org/issues/5203).

This is a follow-up of #7433 which incorporates two Redmine tickets together, one for Suricata secondary mode support and the other for adding an app that would run as a primary process. The PR extends the work of #7283 by continuing in reaching the prefilter concept. As of now, the prefilter configures the provided NICs according to the YAML configuration and then passes packets to Suricata through DPDK rings. It also uses a message system between Suricata and Prefilter for the asynchronous bypass method. Suricata can also load Prefilter configuration by reading Prefilter's shared configuration stored in a memory zone. DPDK Prefilter is intended to be architecturally flexible (not depending on any hardcoded module) but as of now, for some use-cases, DPDK Prefilter uses the Suricata library. 
Use flags `--enable-dpdk --enable-dpdk-apps` to enable Prefilter in the configuration process.
The plan is to have a shared DPDK configuration codebase for DPDK Prefilter and Suricata. As of result, prior to the compilation of the prefilter app, Suricata needs to be first compiled and installed (together with headers and library). Makefile target `dpdk-apps-prereq` should solve the issue. After the Suricata library is installed, it needs to be added to the list of shared libraries (ldconfig) or at least edit `LD_LIBRARY_PATH` variable. Makefile target `dpdk-apps` compiles the prefitler app. The final `prefilter` binary can be found in `dpdk/prefilter/build` folder. 

Changelog from #7433 : 
- updated asynchronous bypass support through message mechanism
- bug fixes
- rebase on the main branch

### CODE REVIEW

File: doc/userguide/configuration/suricata-yaml.rst [hunk 1]
```diff
@@ -2035,6 +2048,83 @@ To be able to run DPDK on Intel cards, it is required to change the default Inte
...
+
+Secondary applications typically do not operate with ports directly but instead, attempt to attach to DPDK rings that should be created by the primary application. 
+Suricata support arbitrary name of the receiving / transmitting rings. 
+The name of the rings specifies the `- interface:` list entry or the `copy-iface` entry. 
+As each Suricata worker reads incoming packets from a unique ring, the ring name contains static and dynamic part.
+Dynamic part, designated as `$QQQ` must be included in the ring name format. The remaining part of the ring name format is static.
+On Suricata startup, workers use the ring name format to lookup DPDK rings. 
+During this process, the dynamic part is substituted with worker's corresponding queue number (starting at 0, going upwards up to the count of Suricata worker threads). 
+
+Suricata in secondary mode does not take in respect configuration entries related to NIC setup (e.g. `mempool-size`, `mempool-cache`).
```
Comment (jufajardini) [inline doc/userguide/configuration/suricata-yaml.rst:2072]: s/ take in respect/ considers

File: doc/userguide/configuration/suricata-yaml.rst [hunk 2]
```diff
@@ -2035,6 +2048,83 @@ To be able to run DPDK on Intel cards, it is required to change the default Inte
...
+.. _dpdk-secondary-ids:
+.. figure:: suricata-yaml/dpdk-secondary-ids.png
+    :align: center
+    :alt: Suricata as a secondary process in IDS mode
+    :figclass: align-center
+
+    `Suricata as a secondary process in IDS mode`
+
+As mentioned, Suricata supports IPS mode even when running as a secondary process. 
+The basic concept of how Suricata attaches to the rings is depicted at Figure :numref:`dpdk-secondary-ips-full`.
```
Comment (jufajardini) [inline doc/userguide/configuration/suricata-yaml.rst:2090]: "how Suricata attaches to the ring" -> this part is a bit unclear. Attaches itself?

Comment (jufajardini) [reply doc/userguide/configuration/suricata-yaml.rst:2090]: s/ at Figure/ in Figure

File: doc/userguide/configuration/suricata-yaml.rst [hunk 3]
```diff
@@ -2035,6 +2048,83 @@ To be able to run DPDK on Intel cards, it is required to change the default Inte
...
+Primary application creates a pair of rings for each NIC it works with.
```
Comment (jufajardini) [inline doc/userguide/configuration/suricata-yaml.rst:2091]: "The primary application..."

File: doc/userguide/configuration/suricata-yaml.rst [hunk 4]
```diff
@@ -2035,6 +2048,83 @@ To be able to run DPDK on Intel cards, it is required to change the default Inte
...
+One ring is used for sending packets to Suricata and the other can be used to retrieve and transmit analyzed packets from Suricata through the appropriate NIC.
+However, in a classic IPS mode, there should be a pair of NICs. Entry `copy-iface` must hold ring name format of the other NIC.
```
Comment (jufajardini) [inline doc/userguide/configuration/suricata-yaml.rst:2093]: Can you reword this a bit? I think the second sentence is hard to understand, as is

File: doc/userguide/configuration/suricata-yaml.rst [hunk 5]
```diff
@@ -2035,6 +2048,83 @@ To be able to run DPDK on Intel cards, it is required to change the default Inte
...
+.. _dpdk-secondary-ips-local:
+.. figure:: suricata-yaml/dpdk-secondary-ips-local.png
+    :align: center
+    :alt: Suricata as a secondary process in local IPS mode
+    :figclass: align-center
+
+    `Suricata as a secondary process in local IPS mode`
+
+
+Pipeline IPS mode (Figure :numref:`dpdk-secondary-ips-pipeline`) extends the previous use-case by chaining multiple applications one after another.
```
Comment (jufajardini) [inline doc/userguide/configuration/suricata-yaml.rst:2116]: "The pipeline IPS... "

File: doc/userguide/configuration/suricata-yaml.rst [hunk 6]
```diff
@@ -2035,6 +2048,83 @@ To be able to run DPDK on Intel cards, it is required to change the default Inte
...
+    :align: center
+    :alt: Suricata as a secondary process in local IPS mode
+    :figclass: align-center
+
+    `Suricata as a secondary process in local IPS mode`
+
+
+Pipeline IPS mode (Figure :numref:`dpdk-secondary-ips-pipeline`) extends the previous use-case by chaining multiple applications one after another. 
+Suricata could receive packets from rings of one application but enqueue them to a ring leading to a different application. 
+The last application has the responsibility to free the packets the mbufs.
```
Comment (jufajardini) [inline doc/userguide/configuration/suricata-yaml.rst:2118]: "to free the packets the mbufs." -> Is something missing here?

**Review by jufajardini — COMMENTED** [review]
Left some suggestions for the documentation page :P

**Comment by lukashino** [comment]
> Left some suggestions for the documentation page :P

@jufajardini made the changes, thanks! They will be part of the next PR.. ;)

**Comment by suricata-qa** [comment]
ERROR:
 

ERROR: QA failed on build_asan.
  

Pipeline 10974

**Comment by suricata-qa** [comment]
ERROR:
 

ERROR: QA failed on build_asan.
  

Pipeline 10974

**Comment by muttanna2972** [comment]
Is there another equivalent PR that is merged?

**Comment by lukashino** [comment]
No, I paused the development at the moment.

**Comment by muttanna2972** [comment]
Thanks @lukashino . In my case I have a primary process (my datapath) reading the packets from NIC. Now can I use the subset of the changes in this PR, to make Suricate a secondary process and read packets from rte_ring shared between my data path and the Suricata?

**Comment by lukashino** [comment]
That's correct, that's exactly as this was designed. Similarly, Suricata then outputted the processed packets to TX ring. So, especially in IPS mode Suricata expects to have a pair rings.
If I recall correctly, "- interface: <IFACENAME>" was put into a ring name template {rx/tx}_IFACENAME_QUEUENUM

**Comment by muttanna2972** [comment]
great, Thank you @lukashino

**Pull request closed without merging**